arangod/Pregel/Algos/AIR/AbstractAccumulator.h

```diff
@@ -0,0 +1,193 @@
+/// @author Lars Maier
+/// @author Markus Pfeiffer
+///
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGODB_PREGEL_ALGOS_VERTEX_ACCUMULATORS_ABSTRACT_ACCUMULATOR_H
+#define ARANGODB_PREGEL_ALGOS_VERTEX_ACCUMULATORS_ABSTRACT_ACCUMULATOR_H 1
+#include <Pregel/Algos/AIR/Greenspun/EvalResult.h>
+#include <iostream>
+#include <numeric>
+
+#include "velocypack/Builder.h"
+#include "velocypack/velocypack-aliases.h"
+
+#include "AccumulatorOptionsDeserializer.h"
+#include "MessageData.h"
+
+namespace arangodb {
+namespace pregel {
+namespace algos {
+namespace accumulators {
+class VertexData;
+
+template <typename T>
+class Accumulator;
+
+struct AccumulatorBase {
+  AccumulatorBase(){};
+  virtual ~AccumulatorBase() = default;
+  template <typename T>
+  Accumulator<T>* castAccumulatorType() {
+    return dynamic_cast<Accumulator<T>*>(this);
+  }
+
+  enum class UpdateResult {
+    CHANGED,
+    NO_CHANGE,
+  };
+
+  // Resets the accumulator to a well-known value
+  virtual auto clear() -> greenspun::EvalResult = 0;
+  virtual auto setBySlice(VPackSlice v) -> greenspun::EvalResult = 0;
+  virtual auto getIntoBuilder(VPackBuilder& result) -> greenspun::EvalResult = 0;
+
+  // This conflates two operations: updating the accumulator, and passing the
+  // sender of the update message into the accumulator, i.e. the message passing and
+  // the accumulator operation
+  // One of these two operations should also be obsolete. This will need some consideration
+  // wrt "efficiency" (whether velocypack is the best format for messages here. It probably is,
+  // in particular since we can prevent copying stuff around)
+  virtual auto updateByMessageSlice(VPackSlice msg) -> greenspun::EvalResultT<UpdateResult> = 0;
+  virtual auto updateByMessage(MessageData const& msg) -> greenspun::EvalResultT<UpdateResult> = 0;
+
+  // used to set state on WorkerContext from message by MasterContext
+  virtual auto setStateBySlice(VPackSlice msg) -> greenspun::EvalResult = 0;
+  // used to set state on WorkerContext from message by MasterContext
+  virtual auto getStateIntoBuilder(VPackBuilder& result) -> greenspun::EvalResult = 0;
+  // used to send updates from WorkerContext to MasterContext, output of this
+  // is given to aggregateStateBySlice on MasterContext
+  virtual auto getStateUpdateIntoBuilder(VPackBuilder& result) -> greenspun::EvalResult = 0;
+  // used to aggregate states on MasterContext after receiving messages from WorkerContexts.
+  virtual auto aggregateStateBySlice(VPackSlice msg) -> greenspun::EvalResult = 0;
+
+  // Is this obsolete?
+  virtual auto finalizeIntoBuilder(VPackBuilder& result) -> greenspun::EvalResult = 0;
+};
+
+template<typename T>
+constexpr auto always_false_v = false;
+
+template <typename T>
+class Accumulator : public AccumulatorBase {
+ public:
+  using data_type = T;
+
+  explicit Accumulator(AccumulatorOptions const&, CustomAccumulatorDefinitions const&){};
```

Comment:
```suggestion
  explicit Accumulator(AccumulatorOptions const&, CustomAccumulatorDefinitions const&) = default;
```

Reply:
This is not a constructor that can be defaulted. @maierlars could you roll this one back?

Reply:
:already_fixed: